### PR TITLE
core: add Hasher and HashableModule

### PR DESCRIPTION
**Pull request by superlopuh**
These utils are planned to be used in `xdsl-gui`, but may well be useful outside of that project. The Hasher contains logic I've used a lot before, it seems to work well enough. In the future, we might want a more efficient way to compute equality/inequality for modules, but we might as well add a working helper class now and optimise later.

### CODE REVIEW

File: tests/utils/test_hashable_module.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from xdsl.dialects.test import TestOp
+from xdsl.utils.hashable_module import HashableModule
+
+
+def m(labels: Iterable[str]):
+    module = ModuleOp([])
+    with ImplicitBuilder(module.body):
+        for label in labels:
+            TestOp(attributes={"label": StringAttr(label)})
+    return module
```
Comment (AntonLydike) [inline tests/utils/test_hashable_module.py:14]: I think having a more descriptive name here would be helpful

**Review by AntonLydike — COMMENTED** [review]
I feel that this is quite a convoluted way of implementing a `__hash__`.

What would be the downside of computing a key (tuple of operation names) and calling `hash` on that? I feel that this might result in much simpler code. (It would also introduce less new code that needs testing)

Furthermore, we should take into account that hashes of mutable objects may result in unexpected/broken behavior, I think having a clearer warning on the HashableModule class would be helpful in communicating the problems associated with that.

**Comment by superlopuh** [comment]
I feel like it's pretty standard, it lets users create a hash of complex objects without having all of the things that compose the hash in memory at a time, making it much more efficient than a hash of a tuple of ops. I also don't feel like it's that complicated, I guess that depends on how used to it the reader is. Swift's hash works like this, so I guess I'm used to it. I also didn't want to go just with xor since that drops the order of elements, but I guess few rewrites just reorder operations, so maybe that should be fine for the sake of hashing in the short term.

**Review by dshaaban01 — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
Nice!
I wouldn't use just a xor here, as it will have a lot of collisions (to be exact, if you have the same parity of every operation, then it will have the same hash, the order won't matter at all).
But that's something we can always fix later, so I'm fine with merging it now!

**Comment by AntonLydike** [comment]
> I wouldn't use just a xor here, as it will have a lot of collisions (to be exact, if you have the same parity of every operation, then it will have the same hash, the order won't matter at all).

I totally agree, we can replace the xor with `hash((self.hash, other_hash))` for it to fall back to a known good hash instead of rolling our own.

**Comment by superlopuh** [comment]
Good idea!

**Pull request merged**